### ISSUE TITLE
Redirects URLs not fully resolving -- MIME-Type being inferred from initial redirect headers

### ISSUE THREAD

**Issue by codingvibe** [savonet/liquidsoap#2517]
**Describe the bug**
When adding URLs to a queue, the URLs are not being fully resolved if they are a redirect. The example I've run in to is an `http://` url that redirects to an `https://` url returns a 301 with a Content-Type of `text/html; charset=UTF-8`, which Liquidsoap isn't a fan of. If the `https://` url is resolved, it returns a 200 with a Content-Type of `audio/mpeg`. This leads me to believe that the requests aren't being redirected appropriately.

**To Reproduce**
```
playlist2 = request.queue(id="myqueue")

def add_track(~protocol,~data,~headers,uri) =
  print("#{data}")
  server.execute("myqueue.push #{data}")
  http.response(
    protocol=protocol,
    code=200,
    content_type="text/html; charset=utf-8",
    data="done"
  )
end

# Harbor HTTP add track function
harbor.http.register(port=8005,method="POST","/track",add_track)
```
Then

```
curl --location --request POST 'http://localhost:8005/track' \
--header 'Content-Type: text/plain' \
--data-raw 'http://some-http-url-that-resolves-to-https/mp3/12345'
```

**Expected behavior**
The URL should be fully resolved before trying to infer MIME-Type.

**Version details**
 - Windows 10
 - 20.06

**Install method**
Running EXE from releases page.

**Comment by codingvibe** [comment]
I tried just using the `https://`, resolved URL in my POST and it's still getting the same error. I'm not sure what the backend implementation of the URL I'm using is, but the return headers suggest it's backed by some Cloudflare hosting. When in Postman with redirects off, the `https://` URL resolves, returns the file, and has the right `Content-Type`.

I'm running my liquidsoap server locally from my machine. Could it be an SSL thing and the request is failing in a way that liquidsoap is still trying to process correctly?

**Comment by codingvibe** [comment]
Another update: when I override `http_protocol` to call `http.get` instead of `http.head` with this code:

```  
  ret = http.get(timeout_ms=int_of_float(timeout*1000.), redirect=true, uri)
  code = ret.status_code ?? 999
  headers = ret.headers

  log(level=3,"got code #{code}")
  log(level=3,"#{ret}")
```
I get this output:

```
2022/07/18 22:50:34 [procol.external:3] got code 200
2022/07/18 22:50:34 [procol.external:3] MP3 no set
```

and If I leave it as `http.head` printing `ret` returns a much richer response (redacted sensitive info):

```
 {headers = [("alt-svc", "h3=\":443\"; ma=86400, h3-29=\":443\"; ma=86400"), ("server", "cloudflare"), ("nel", "{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}"),  ("cf-cache-status", "DYNAMIC"), ("connection", "keep-alive"), ("content-type", "text/html; charset=UTF-8"), ("date", "Tue, 19 Jul 2022 03:54:21 GMT")], status_message = "OK", status_code = 200, protocol_version = "1.1"}
```

Is this some issue with the API I'm using or something in Liquidsoap? Now I'm not so sure lol

**Comment by toots** [comment]
Thanks for the report. Could you send the (redacted) result of `curl -v <url>` for the redirect `http` url and also the final `https` url?

**Comment by toots** [comment]
Ho, I forgot: do you have any log around the corresponding fetch?

**Comment by codingvibe** [comment]
The default logs looked like this when failing the fetch:

```
2022/07/16 14:29:09 [request.dynamic:3] Fetch failed: retry.
http://<my_source_url>/<a_php_file_for_some_reason>.php/?mp3=<fileId>
2022/07/16 14:32:07 [procol.external:3] No known file extension for mime: text/html; charset=UTF-8
2022/07/16 14:32:08 [decoder:3] No decoder available for "C:\\Users\\user\\AppData\\Local\\Temp\\liq-processceed44.osb"
```

Here's the curl output for the direct `https` call

```
$ curl -v https://redacted.redacted.redacted/serve.php/?mp3=1234
*   Trying 172.67.69.105:443...
* TCP_NODELAY set
* Connected to redacted.redacted.redacted (172.67.69.105) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jun  6 00:00:00 2022 GMT
*  expire date: Jun  5 23:59:59 2023 GMT
*  subjectAltName: host "redacted.redacted.redacted" matched cert's "*.redacted.redacted"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x56122a2e17c0)
> GET /serve.php/?mp3=1234 HTTP/2
> Host: redacted.redacted.redacted
> user-agent: curl/7.68.0
> accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
< HTTP/2 200
< date: Wed, 20 Jul 2022 14:23:38 GMT
< content-type: audio/mpeg
< content-length: 3693089
< accept-ranges: bytes
< content-disposition: attachment; filename="1234.mp3"
< expires: Tue, 18 Oct 2022 16:23:38 GMT
< cf-cache-status: DYNAMIC
< expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=BJ94kt%2FsGgeDdL4aq52OBy0M%2FtvxWo%2BoLTsdYmsV7xvtweRSOvqjjkOJJWEw6hRQ%2BCcPmEuxST6j9Z85XBLSueL7wevKcKwyYEFylQNhwHlAFZzoeOa63NBL8NmG24KLpOM%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 72dc5b177f947da6-LAX
< alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
<
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.
* Failed writing body (0 != 810)
* stopped the pause stream!
* Connection #0 to host redacted.redacted.redacted left intact
```

and for the http:

```
$ curl -v http://redacted.redacted.redacted/serve.php/?mp3=1234
*   Trying 104.26.8.42:80...
* TCP_NODELAY set
* Connected to redacted.redacted.redacted (104.26.8.42) port 80 (#0)
> GET /serve.php/?mp3=1234 HTTP/1.1
> Host: redacted.redacted.redacted
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Date: Wed, 20 Jul 2022 14:27:49 GMT
< Content-Type: text/html; charset=iso-8859-1
< Transfer-Encoding: chunked
< Connection: keep-alive
< Location: https://redacted.redacted.redacted/serve.php/?mp3=1234
< CF-Cache-Status: DYNAMIC
< Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=%2FG0cleasFFO2Vn6YK3PlX4F92Rvi2ZnpDutMu0oGpzqXiYALm3dEOuu9XqxjKwm9qovx6BsRedEUzk3dKR4h0s3Y0wlDcH0cKyaj05skxML%2BuQagiqsNHivCnvyrr98fsi8%3D"}],"group":"cf-nel","max_age":604800}
< NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< Server: cloudflare
< CF-RAY: 72dc613a798b7ec3-LAX
< alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
<
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="https://redacted.redacted.redacted/serve.php/?mp3=1234">here</a>.</p>
<hr>
<address>Apache/2.4.29 (Ubuntu) Server at redacted.redacted.redacted Port 80</address>
</body></html>
* Connection #0 to host redacted.redacted.redacted left intact
```

Yeah, the `https` call resolves to the right file using curl, but for some reason, when liquidsoap is doing the call, it returns a body that says  `MP3 no set`, which is very strange.

**Comment by toots** [comment]
Interesting! Have you tried printing the `uri` that is passed from your parameters to the `http.get` call?

**Comment by codingvibe** [comment]
Sure! I redefined `http_protocol` like this:

```
%ifdef http.head
# Resolve http(s) URLs using curl
# @flag hidden
def http_protocol(proto,~rlog,~maxtime,arg) =
  uri = "#{proto}:#{arg}"

  def log(~level,s) =
    rlog(s)
    log(label="procol.external",level=level,s)
  end

  timeout = maxtime - time()
  log(level=3,"uri: #{uri}")

  ret = http.get(timeout_ms=int_of_float(timeout*1000.), redirect=true, uri)
  code = ret.status_code ?? 999
  headers = ret.headers

  log(level=3,"got code #{code}")
  log(level=3,"content type #{headers}")
  log(level=3,"#{ret}")
  log(level=3,"break here")

  mime =
    if 200 <= code and code < 300 then
      headers["content-type"]
    else
      log(level=3,"Failed to fetch mime-type for #{uri}.")
      log(level=4,"Request response: #{ret}")
      null()
    end

  extname = null.case(mime, (fun () -> "osb"), (fun (mime) ->
    if list.mem(mime, ["audio/mpeg", "audio/mp3"]) then
      "mp3"
    elsif list.mem(mime,["application/ogg", "application/x-ogg",
                         "audio/x-ogg", "audio/ogg", "video/ogg"]) then
      "ogg"
    elsif list.mem(mime, ["audio/flac", "audio/x-flac"]) then
      "flac"
    elsif list.mem(mime,["audio/mp4", "application/mp4", "video/mp4"]) then
      "mp4"
    elsif list.mem(mime,["audio/vnd.wave", "audio/wav",
                         "audio/wave", "audio/x-wav"]) then
      "wav"
    else
      log(level=3,"No known file extension for mime: #{mime}")
      "osb"
    end))

  output = file.temp("liq-process", ".#{extname}")

  file_writer = file.write.stream(output)

  timeout = maxtime - time()
  
  response = http.get.stream(
    on_body_data=file_writer,
    timeout_ms=int_of_float(timeout * 1000.),
    uri
  )

  if response.status_code < 400 then
    [output]
  else
    log(level=3, "Error while fetching http data: #{response.status_code} - #{response.status_message}")
    []
  end
end
```

and got the following output:
```
2022/07/23 13:35:55 [procol.external:3] uri: https://redacted.redacted.redacted/serve.php/?mp3=15236
2022/07/23 13:35:56 [procol.external:3] got code 200
2022/07/23 13:35:56 [procol.external:3] content type [("alt-svc", "h3=\":443\"; ma=86400, h3-29=\":443\"; ma=86400"), ("cf-ray", "72f684a2df3d7bc7-LAX"), ("server", "cloudflare"), ("nel", "{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}"), ("report-to", "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=kRMIdCLDrpLnU4AOTJ3k3IEO30ilDVyr6%2B%2B2ulkp4ALPnh5nc1DNxbgNOpOkc793rKfjlHXAMOnlFhGAydIUdpKz%2FyGnfQWu%2F%2FVR5pHuRHnZt%2FulHmTGFEIOl342EbsDZgI%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}"), ("expect-ct", "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""), ("cf-cache-status", "DYNAMIC"), ("connection", "keep-alive"), ("transfer-encoding", "chunked"), ("content-type", "text/html; charset=UTF-8"), ("date", "Sat, 23 Jul 2022 18:35:50 GMT")]
2022/07/23 13:35:56 [procol.external:3] MP3 no set
2022/07/23 13:35:56 [procol.external:3] break here
2022/07/23 13:35:56 [procol.external:3] No known file extension for mime: text/html; charset=UTF-8
```

**Comment by toots** [comment]
Thanks!

Could you try with the windows build here: https://github.com/savonet/liquidsoap/actions/runs/2724686474 ? This should tell use what url is effectively being used internally for the redirect..

**Comment by codingvibe** [comment]
I set the log level to 5 and fished these out of the output:

```
2022/07/23 14:56:05 [harbor:4] New client on port 8005: 127.0.0.1
2022/07/23 14:56:05 [myserver:5] Queue is empty!
2022/07/23 14:56:05 [myserver:5] Failed to prepare track: no file.
2022/07/23 14:56:05 [harbor:4] Method: POST, uri: /track, protocol: HTTP/1.1
2022/07/23 14:56:05 [harbor:4] Header: Content-Type, value: text/plain.
2022/07/23 14:56:05 [harbor:4] Header: User-Agent, value: PostmanRuntime/7.29.2.
2022/07/23 14:56:05 [harbor:4] Header: Accept, value: */*.
2022/07/23 14:56:05 [harbor:4] Header: Postman-Token, value: b0c5ab24-878d-4efc-a419-d3f503ba6976.
2022/07/23 14:56:05 [harbor:4] Header: Host, value: localhost:8005.
2022/07/23 14:56:05 [harbor:4] Header: Accept-Encoding, value: gzip, deflate, br.
2022/07/23 14:56:05 [harbor:4] Header: Connection, value: keep-alive.
2022/07/23 14:56:05 [harbor:4] Header: Content-Length, value: 45.
2022/07/23 14:56:05 [harbor:4] HTTP POST request on /track.
https://redacted.redacted.redacted/serve.php/?mp3=15236
2022/07/23 14:56:05 [harbor:4] Found handler 'POST /track' on port 8005.
2022/07/23 14:56:05 [myserver:4] Pushing <request(id=0)> on the queue.
2022/07/23 14:56:05 [request:5] Resolving request [[https://redacted.redacted.redacted/serve.php/?mp3=15236]].
2022/07/23 14:56:05 [procol.external:3] uri: https://redacted.redacted.redacted/serve.php/?mp3=15236
2022/07/23 14:56:05 [myserver:5] Queue is empty!
2022/07/23 14:56:06 [procol.external:3] got code 200
2022/07/23 14:56:06 [procol.external:3] content type [("alt-svc", "h3=\":443\"; ma=86400, h3-29=\":443\"; ma=86400"), ("cf-ray", "72f6fa0ff86c7ba1-LAX"), ("server", "cloudflare"), ("nel", "{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}"), ("report-to", "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=K%2FZwIVpqZbYQpU7kSjPJgFpax0Ax5BQDKD3HHPsr7MiGV57HOz8gl8nnsNv1YXzAIJoREnRsU8CjfSNPQknqDiZpGVjf51wQJtIWwFFLKq9D%2BGXwz27Lt6bHh6DhbK2JHsI%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}"), ("expect-ct", "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""), ("cf-cache-status", "DYNAMIC"), ("connection", "keep-alive"), ("transfer-encoding", "chunked"), ("content-type", "text/html; charset=UTF-8"), ("date", "Sat, 23 Jul 2022 19:56:00 GMT")]
2022/07/23 14:56:06 [procol.external:3] MP3 no set
2022/07/23 14:56:06 [procol.external:3] break here
2022/07/23 14:56:06 [procol.external:3] No known file extension for mime: text/html; charset=UTF-8
2022/07/23 14:56:07 [decoder.ffmpeg:4] Unsupported file extension for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"!
2022/07/23 14:56:07 [decoder.ogg:4] Unsupported file extension for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"!
2022/07/23 14:56:07 [decoder.srt:4] Unsupported file extension for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"!
2022/07/23 14:56:07 [decoder.aiff:4] Unsupported file extension for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"!
2022/07/23 14:56:07 [decoder.wav:4] Unsupported file extension for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"!
2022/07/23 14:56:07 [decoder:4] Available decoders: IMAGE (priority: 1)
2022/07/23 14:56:07 [decoder:4] Trying decoder "IMAGE"
2022/07/23 14:56:07 [decoder:4] Trying method "FFMPEG" for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"...
2022/07/23 14:56:07 [decoder:4] Trying method "PPM" for "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb"...
2022/07/23 14:56:07 [decoder:4] Decoder "PPM" failed on "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb": Mm_image.ImageRGBA32.Invalid_format("Not a PPM file.")!
2022/07/23 14:56:07 [decoder:3] Unable to decode "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb" using image decoder(s)!
2022/07/23 14:56:07 [decoder:3] Available decoders cannot decode "C:\\Users\\colby\\AppData\\Local\\Temp\\liq-process7542f9.osb" as {audio=pcm(stereo),video=none,midi=none}
2022/07/23 14:56:07 [request:5] Resolved to [].
2022/07/23 14:56:07 [request.dynamic:3] Fetch failed: retry.
```

This is preceded and followed by a bunch of:

```
2022/07/23 14:56:05 [myserver:5] Queue is empty!
2022/07/23 14:56:05 [myserver:5] Failed to prepare track: no file.
2022/07/23 14:56:05 [myserver:5] Queue is empty!
2022/07/23 14:56:05 [myserver:5] Failed to prepare track: no file.
```

but I'm pretty sure those are unrelated.

I can also change the title of this ticket if we end up finding out it's because parameters aren't being sent/parameterized correctly so future Googlers will be set down the right path

**Comment by toots** [comment]
Thanks. The log is a `printf` so it might appear somewhere different. It should read: `Redirecting to: "https:..."`

**Comment by codingvibe** [comment]
It looks like the redirect is working. The print statements look like this:

```
2022/07/24 11:06:48 [procol.external:3] uri: http://redacted.redacted.redacted/serve.php/?mp3=15236
Redirecting to: "https://redacted.redacted.redacted/serve.php/?mp3=15236"
```

I think the actual problem is the query params aren't being sent correctly. When I hit the URL without any query params with Postman, the body that's returned is `MP3 no set`, which is what liquidsoap is getting

**Comment by toots** [comment]
Cool thanks. Yes, I agree with your assumption I wanted to see what was hitting the curl API. Is there any chance that you could share the full url in private? I'm at toots@rastageeks.org.

**Comment by toots** [comment]
Ok, I was able to trace the error back to the `ocurl` minimal example and found out that it was due to the `User-Agent` not being set.

This works:
```
> GET /serve.php/?mp3=15236 HTTP/2
> Host: mp3.chillhop.com
> user-agent: curl/7.77.0
> accept: */*
```

But this does not:
```
> GET /serve.php/?mp3=15236 HTTP/2
> Host: mp3.chillhop.com
> accept: */*
```

The issues is now fixed in `rolling-release-v2.1.x` and `main` here. It was a good call to add the user-agent. However, I believe that is technically an issue with the remote HTTP server as both requests should be valid ones.

**Comment by toots** [comment]
Ok, I was able to trace the error back to the `ocurl` minimal example and found out that it was due to the `User-Agent` not being set.

This works:
```
> GET /serve.php/?mp3=15236 HTTP/2
> Host: redacted
> user-agent: curl/7.77.0
> accept: */*
```

But this does not:
```
> GET /serve.php/?mp3=15236 HTTP/2
> Host: redacted
> accept: */*
```

The issue is now fixed in `rolling-release-v2.1.x` and `main` here. It was a good call to add the user-agent. However, I believe that is technically an issue with the remote HTTP server as both requests should be valid ones.
